src/ore/src/iter.rs

```diff
@@ -86,6 +107,37 @@ impl<I: Iterator> Iterator for ExactSize<I> {
 
 impl<I: Iterator> ExactSizeIterator for ExactSize<I> {}
 
+/// Iterator type returned by [`IteratorExt::repeat_clone`].
+#[derive(Debug)]
+pub struct RepeatClone<I: Iterator, A> {
+    iter: I,
+    next: Option<I::Item>,
+    extra_val: Option<A>,
+}
+
+impl<I: Iterator, A: Clone> Iterator for RepeatClone<I, A> {
+    type Item = (I::Item, A);
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
If you store a `Peekable` iterator you can simplify this logic and avoid calling `next` on construction:

```rust

impl<I: Iterator, A: Clone> Iterator for RepeatClone<I, A> {
    type Item = (I::Item, A);

    fn next(&mut self) -> Option<Self::Item> {
        let next = self.iter.next()?;

        // Clone the extra_val only if there is an item to return on the next call to `next`.
        let val = match self.iter.peek() {
            Some(_) => self.extra_val.clone(),
            None => self.extra_val.take(),
        };

        Some((next, val.expect("RepeatClone invariant violated")))
    }
}

impl<I: Iterator<Item: Debug> + Debug, A: Debug> Debug for RepeatClone<I, A> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RepeatClone")
            .field("iter", &self.iter)
            .field("extra_val", &self.extra_val)
            .finish()
    }
}
```